content/blog/2019-09-25-bril2jb.md

```diff
@@ -0,0 +1,75 @@
++++
+title = "bril2jb: A tool that translates bril to Java Bytecode"
+extra.author = "Siqiu Yao"
+extra.bio = """
+  [Siqiu Yao](http://www.cs.cornell.edu/~yaosiqiu/) is a 3rd year Ph.D. student interested in security about systems and programming languages.~~~~~~.
+"""
++++
+
+### Goal
+The goal of this project is to provide a tool to translate bril code to Java Bytecode.
+
+### Design
+#### Before the actual translation process
+At first glance, translating bril to [Java Bytecode](https://en.wikipedia.org/wiki/Java_bytecode) seems a 
+pretty straightforward job, since bril (currently) only contains simple instructions and supports no function invocation.
+But it turns out that making something directly runnable by *Java Virtual Machine*(JVM, the run-time environment of Java Bytecode)
+is not trivial, because we need to construct our program as a Java *class*, 
+which is the only code format JVM accepts.
+
+Therefore we need to create a wrapper class(say `Wrapper.class`) like the following :
+```JAVA
+public class Wrapper {
+  public Wrapper();
+  public static void main(java.lang.String[]);
+}
+```
+And the `main` function would contain the instructions in the bril code.
+
+#### The translation process
+I chose [ASM](https://asm.ow2.io), a Java Bytecode manipulation and analysis framework to help with the translation.
+
+The translation process is pretty regular. One interesting point is that Java Bytecode is strictly-typed, 
+while bril is naturally dynamic and 
+there might be no explicit declaration of a variable before it is used(in the order of instruction list),
+and some instructions are polymorphic such as `print`.
```

Comment:
I disagree that Bril is “naturally dynamic.” There are no variable declarations, it’s true, but it’s statically and strictly typed.

This would perhaps be a good place to discuss your design decision (if you made one) about how to deal with programs that use different types for the same variable name. I know this came up on Slack, but the world doesn’t know that. :)

Reply:
You say “some instructions” are polymorphic, but is anything other that “print” actually polymorphic? Please be precise about this.